### PR TITLE
add requester_pays boolean argument to all S3 requests

### PR DESCRIPTION
**Pull request by barko**

### CODE REVIEW

File: aws-s3/s3.ml [hunk 1]
```diff
@@ -231,7 +231,13 @@ module Make(Io : Types.Io) = struct
   type range = { first: int option; last: int option }
 
   type nonrec 'a result = ('a, error) result Deferred.t
-  type 'a command = ?credentials:Credentials.t -> ?connect_timeout_ms:int -> endpoint:Region.endpoint -> 'a
+  type 'a command = ?credentials:Credentials.t -> ?connect_timeout_ms:int -> ?requester_pays:bool -> endpoint:Region.endpoint -> 'a
+
+  (* conditionally add a header indicating caller's willingness to pay
+     for AWS data transfer costs *)
+  let add_request_payer headers = function
```
Comment (andersfugmann) [inline aws-s3/s3.ml:238]: As I understand the API, this header must be present on calls to buckets where the requested acknowledges that the requester will be charged for the request. 

The name `request_payer` seems not to reflect this. I suggest renaming this to `confirm_requester_pays`, in line with the documentation.

The function here acually only adds the header if the second argument is true, so it should be named `maybe_...`. Also flipping the two arguments makes it easier to chain calls using the pipe (`|>`) operator.

File: aws-s3/s3.ml [hunk 2]
```diff
@@ -270,7 +276,8 @@ module Make(Io : Types.Io) = struct
       let resp = Error_response.of_xmlm_exn (xmlm_of_string body) in
       Deferred.return (Error (Unknown (code, resp.code)))
 
-  let put_common ?credentials ?connect_timeout_ms ~endpoint ?content_type ?content_encoding ?acl ?cache_control ?expect ~bucket ~key ~body () =
+
+  let put_common ?credentials ?connect_timeout_ms ?(requester_pays=false) ~endpoint ?content_type ?content_encoding ?acl ?cache_control ?expect ~bucket ~key ~body () =
```
Comment (andersfugmann) [inline aws-s3/s3.ml:280]: As mentioned before, the argument name is misleading.

File: aws-s3/s3.mli [hunk 3]
```diff
@@ -36,7 +36,9 @@ module Make(Io : Types.Io) : sig
   }
 
   type nonrec 'a result = ('a, error) result Deferred.t
-  type 'a command = ?credentials:Credentials.t -> ?connect_timeout_ms:int -> endpoint:Region.endpoint -> 'a
+
+  (* If [requester_pays], caller will pay AWS data transfer costs. *)
```
Comment (andersfugmann) [inline aws-s3/s3.mli:40]: This is not in line with the documentation. When operating on `Requester Pays buckets`, a special header must be present to indicate the callers willingness to accept charges. Setting the flag on non-requester pays bucket has no effect, and not setting the flag on requester-pays buckets, wil case any operation to fail (unless the caller owns the bucket).

File: cli/cli.ml [hunk 4]
```diff
@@ -33,17 +33,23 @@ let parse exec =
 
   let retries =
     let doc = "Retries in case of error" in
-    Arg.(value & opt int 0 & info ["retries"] ~docv:"RETIES" ~doc)
+    Arg.(value & opt int 0 & info ["retries"] ~docv:"RETRIES" ~doc)
```
Comment (andersfugmann) [inline cli/cli.ml:36]: Nice catch. Thanks

File: cli/cli.ml [hunk 5]
```diff
@@ -33,17 +33,23 @@ let parse exec =
...
   in
 
   let expect =
     let doc = "Use expect -> 100-continue for put/upload_chunk" in
     Arg.(value & flag & info ["expect"; "e"] ~docv:"EXPECT" ~doc)
   in
 
+  let requester_pays =
+    let doc = "indicate that the client is paying for the request" in
```
Comment (andersfugmann) [inline cli/cli.ml:45]: Again, I think this is not in line with what this flag actually means.

File: aws-s3/s3.mli (hunk 3 above)
Comment (andersfugmann) [reply aws-s3/s3.mli:40]: Maybe add a link to the [documentation](https://docs.aws.amazon.com/AmazonS3/latest/userguide/ObjectsinRequesterPaysBuckets.html) in the documentation comment for the flag.

**Review by andersfugmann — CHANGES_REQUESTED** [review]
Thanks for the PR.
Some comments on the naming on the flag and extending the documentation but otherwise ok.

Thanks for submitting.

**Comment by andersfugmann** [comment]
The PR seems to have conflicts after recent changes. Would you be able to rebase on master?

**Comment by barko** [comment]
Thanks for the many suggestions. I believe that I adopted all of them in new PR #27 .

**Pull request closed without merging**